File: python/problem_094.py
#!/usr/bin/python
# 
# It is easily proved that no equilateral triangle exists with integral length
# sides and integral area. However, the almost equilateral triangle 5-5-6 has
# an area of 12 square units.
# 
# We shall define an almost equilateral triangle to be a triangle for which two
# sides are equal and the third differs by no more than one unit.
# 
# Find the sum of the perimeters of all almost equilateral triangles with
# integral side lengths and area and whose perimeters do not exceed one billion
# (1,000,000,000).
# 
# http://projecteuler.net/index.php?section=problems&id=94

import math, euler

def heron(a, b, c):
    s = (a + b + c) / 2
    return math.sqrt(s * (s - a) * (s - b) * (s - c))

def triangles(maximum):
    ab = 2
    while True:
        c = ab - 1
        if 2 * ab + c > maximum:
            raise StopIteration
        yield (ab, ab, c)
        c = ab + 1
        if 2 * ab + c > maximum:
            raise StopIteration
        yield (ab, ab, c)
        ab += 1

def is_integral(n):
    return int(n) == n

def test(t):
    # if t[0] % 1e5 == 0:
    #     print "Testing %s" % str(t)
    return is_integral(heron(t[0], t[1], t[2]))
    # a, b = t[0], t[2]
    # return is_integral(math.sqrt(((a*a * b*b) / 4) - ((b*b*b*b) / 16)))

if __name__ == '__main__':
    print len(filter(test, triangles(1e6)))

File: haskell/problem_094.hs
-- It is easily proved that no equilateral triangle exists with integral length
-- sides and integral area. However, the almost equilateral triangle 5-5-6 has
-- an area of 12 square units.
-- 
-- We shall define an almost equilateral triangle to be a triangle for which
-- two sides are equal and the third differs by no more than one unit.
-- 
-- Find the sum of the perimeters of all almost equilateral triangles with
-- integral side lengths and area and whose perimeters do not exceed one
-- billion (1,000,000,000).
-- 
-- http://projecteuler.net/index.php?section=problems&id=94

module Main where

is_int n = f == c where f = floor n
                        c = ceiling n

limit = 10 ** 9

almosts = [floor p | a <- [1..(limit/3)]
                   , b <- [a - 1, a + 1]
                   , let p = 2 * a + b
                   , let area = 0.5 * b * (sqrt $ (a ** 2) - 0.25 * (b ** 2))
                   , is_int area
                   ]

main :: IO ()
main = do mapM_ print almosts
          putStrLn "----------"
          print $ sum almosts

